### PR TITLE
Order model

### PR DESCRIPTION
**Pull request by SaiyaJ1N**

### CODE REVIEW

**Review by Someboty — APPROVED** [review]

File: src/main/java/com/onlinebookshop/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                                                      @PathVariable Long id, Pageable pageable) {
+        return orderService.getOrderItemFromOrder(orderId, id, pageable);
+    }
+
+    @Operation(summary = "Update Order status",
+            description = "Update the status of an order")
+    @PatchMapping("/{id}")
+    @PreAuthorize("hasRole('ADMIN')")
+    public OrderResponseDto updateOrderStatus(@PathVariable Long id,
+                                              @RequestBody UpdateOrderStatusDto updateRequestDto) {
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookshop/controller/OrderController.java]: Will updateRequestDto be validated?

File: src/main/java/com/onlinebookshop/service/impl/OrderServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    @Override
+    @Transactional
+    public OrderResponseDto save(OrderRequestDto requestDto) {
+        Order order = new Order();
+        User user = authenticationService.getCurrentUser();
+        ShoppingCart shoppingCart = authenticationService.getCartOfCurrentUser();
+        order.setUser(user);
+        order.setOrderDate(LocalDateTime.now());
+        order.setStatus(Order.Status.PENDING);
+        order.setTotal(shoppingCart.getCartItems().stream()
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookshop/service/impl/OrderServiceImpl.java]: Please move logic of total order calculation to separate method

File: src/main/java/com/onlinebookshop/service/impl/OrderServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                .map(cartItem -> cartItem.getBook()
+                        .getPrice()
+                        .multiply(BigDecimal.valueOf(cartItem.getQuantity())))
+                .reduce(BigDecimal.valueOf(0), BigDecimal::add));
+        order.setShippingAddress(requestDto.getShippingAddress());
+        Order savedOrder = orderRepository.save(order);
+        savedOrder.setOrderItems(shoppingCart.getCartItems().stream()
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookshop/service/impl/OrderServiceImpl.java]: Is it possible to save orderItems together with order entity? (check out cascades)

File: src/main/java/com/onlinebookshop/service/impl/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                .map(orderItemMapper::toOrderItem)
+                .peek(orderItem -> orderItem.setOrder(savedOrder))
+                .peek(this::saveOrderItemToDb)
+                .collect(Collectors.toSet()));
+        return orderMapper.toResponseDto(savedOrder);
+    }
+
+    @Override
+    public List<OrderItemResponseDto> getAllOrderItemsByOrderId(Long id, Pageable pageable) {
+        Order order = orderRepository.findById(id).orElseThrow(() ->
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookshop/service/impl/OrderServiceImpl.java]: Please make sure that the user can retrieve only their own orders, not order created bu other users

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
GJ! Let`s improve your solution

File: src/main/java/com/onlinebookshop/service/impl/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        return orderMapper.toResponseDto(savedOrder);
+    }
+
+    @Override
+    public List<OrderItemResponseDto> getAllOrderItemsByOrderId(Long id, Pageable pageable,
+                                                                Authentication authentication) {
+        Order order = orderRepository.findById(id).orElseThrow(() ->
+                new NoSuchElementException("Can`t find order by id: " + id));
+        User user = (User) authentication.getPrincipal();
+        if (!order.getUser().getId().equals(user.getId())) {
```
Comment (andrii-hoienko) [inline src/main/java/com/onlinebookshop/service/impl/OrderServiceImpl.java]: 1. You should be sure that all interactions with order should be performed by the user that owns an order.
2. OPTIONAL I would prefer to check that order belongs to a specific user on the DB level (modify query to make it return only order that belongs to user by adding additional where for example)

File: src/main/java/com/onlinebookshop/service/impl/OrderServiceImpl.java (hunk 3 above)
Comment (andrii-hoienko) [reply src/main/java/com/onlinebookshop/service/impl/OrderServiceImpl.java]: Discussed in the private channel

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
GJ! One minor issue left

**Review by andrii-hoienko — APPROVED** [review]
GJ!

**Pull request merged**